feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,567 @@
+	}
+	return policyRule
+}
+
+func getMonKeyFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) []byte {
+	if pccRule.RefUmData != nil && len(*pccRule.RefUmData) > 0 && policyDecision.UmDecs != nil {
+		umData, found := policyDecision.UmDecs.Get((*pccRule.RefUmData)[0])
+		if found {
+			return []byte(umData.UmId)
+		}
+	}
+	return []byte{}
+}
+
+func getTcDataFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.RedirectInformation {
+	if pccRule.RefTcData != nil && len(*pccRule.RefTcData) > 0 && policyDecision.TraffContDecs != nil {
+		tcData, found := policyDecision.TraffContDecs.Get((*pccRule.RefTcData)[0])
+		if found {
+			return GetRedirectInfoProto(tcData.RedirectInfo)
+		}
+	}
+	return nil
+}
+
+func getQosDataFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.FlowQos {
+	if pccRule.RefQosData != nil && len(*pccRule.RefQosData) > 0 && policyDecision.QosDecs != nil {
+		qosData, found := policyDecision.QosDecs.Get((*pccRule.RefQosData)[0])
+		if found {
+			return GetFlowQosProto(&qosData)
+		}
+	}
+	return nil
+}
+
+func getFlowListFromPccRule(pccRule *n7_sbi.PccRule) []*protos.FlowDescription {
+	var flowList []*protos.FlowDescription
+	if pccRule.FlowInfos != nil {
+		for _, flowInfo := range *pccRule.FlowInfos {
+			if flowInfo.FlowDescription != nil {
+				flow, err := policydb.GetFlowDescriptionFromFlowString(string(*flowInfo.FlowDescription))
+				if err != nil {
+					glog.Errorf("Could not get flow for description %s : %s", *flowInfo.FlowDescription, err)
+				} else {
+					flowList = append(flowList, flow)
+				}
+			}
+		}
+	}
+	return flowList
+}
+
+func getTrackingTypeProto(monKeyPresent bool, ratingGrpPresent bool) protos.PolicyRule_TrackingType {
+	if monKeyPresent && ratingGrpPresent {
+		return protos.PolicyRule_OCS_AND_PCRF
+	} else if monKeyPresent && !ratingGrpPresent {
+		return protos.PolicyRule_ONLY_PCRF
+	} else if !monKeyPresent && ratingGrpPresent {
+		return protos.PolicyRule_ONLY_OCS
+	} else {
+		return protos.PolicyRule_NO_TRACKING
+	}
+}
+
+func GetRedirectInfoProto(redirInfo *n7_sbi.RedirectInformation) *protos.RedirectInformation {
+	redirProto := &protos.RedirectInformation{}
+	if redirInfo == nil {
+		return redirProto
+	}
+	if redirInfo.RedirectEnabled != nil {
+		redirProto.Support = protos.RedirectInformation_DISABLED
+		if *redirInfo.RedirectEnabled {
+			redirProto.Support = protos.RedirectInformation_ENABLED
+		}
+	}
+	if redirInfo.RedirectAddressType != nil {
+		redirProto.AddressType = GetRedirectAddrTypeProto(redirInfo.RedirectAddressType)
+	}
+	if redirInfo.RedirectServerAddress != nil {
+		redirProto.ServerAddress = *redirInfo.RedirectServerAddress
+	}
+
+	return redirProto
+}
+
+func GetRedirectAddrTypeProto(
+	redirAddrType *n7_sbi.RedirectAddressType,
+) protos.RedirectInformation_AddressType {
+	if redirAddrType == nil {
+		return protos.RedirectInformation_IPv4
```

Comment:
do you really need to povide redirect information when it is nil?

Note redirect will be empty almost always. Now you are converting this to a specific value. Is that correct?

Reply:
See comment below